src/graph/create-entity.ts

```diff
@@ -0,0 +1,75 @@
+import { publishEdit } from '../ipfs.js';
+import { Triple } from '../triple.js';
+import type { SetTripleOp } from '../types.js';
+import { DEFAULT_API_HOST } from './constants.js';
+
+type Params = {
+  id: string;
+  spaceId: string;
+  accountId: string;
+  data: {
+    [filedName: string]: string;
+  };
+  mapping: {
+    [filedName: string]: string;
+  };
+  options?: {
+    apiHost?: string;
+    network?: 'TESTNET' | 'MAINNET';
+  };
+};
+
+export const createEntity = async ({ id, data, spaceId, accountId, mapping, options }: Params) => {
+  const ops: SetTripleOp[] = [];
+  const apiHost = options?.apiHost || DEFAULT_API_HOST;
+  const network = options?.network || 'TESTNET';
+  for (const [fieldName, value] of Object.entries(data)) {
+    const attributeId = mapping[fieldName];
+    if (!attributeId) {
+      throw new Error(`Attribute ID for field ${fieldName} not found`);
+    }
+    let attributeValueType: 'TEXT' | 'CHECKBOX' | 'NUMBER' | 'URL' | 'POINT';
+    switch (typeof value) {
+      case 'string':
+        attributeValueType = 'TEXT';
+        break;
+      case 'boolean':
+        attributeValueType = 'CHECKBOX';
+        break;
+      case 'number':
+        attributeValueType = 'NUMBER';
+        break;
+      default:
+        throw new Error('Not implemented value type');
+    }
+    const setTripleOp: SetTripleOp = Triple.make({
+      entityId: id,
+      attributeId,
+      value: {
+        type: attributeValueType,
+        value,
+      },
+    });
+
+    ops.push(setTripleOp);
+  }
+
+  const cid = await publishEdit({
+    name: 'Edit name',
+    ops,
+    author: accountId,
+  });
+
+  const result = await fetch(`${apiHost}/space/${spaceId}/edit/calldata`, {
+    method: 'POST',
+    body: JSON.stringify({
+      cid: cid,
+      network,
+    }),
+  });
+
+  const jsonResult = await result.json();
+  return jsonResult;
+
+  // TODO add walletClient.sendTransaction
+};
```

Comment:
For the GRC-20 SDK my expectation is that it composes ops at the lowest level for publishing via an edit. We want developers to group as many ops into an edit as logically possible rather than publishing to IPFS for every change that they make.

If we include the `createEntity` function into the SDK, I'd expect it would just generate the ops representing that entity and just return the ops, rather than publishing them to IPFS + onchain. Creating an entity is still fairly granular and developers probably want to include this created entity into an edit with other ops.

Reply:
I think this is becoming a question around "what's the difference between this sdk and hypergraph?"

Reply:
makes sense, would then split it up into:

- `createEntity` returning the ops
- `publish` accepting a list of ops and doing the actual publishing